[LBSE] Transform update handling is unreliable

nikolaszimmermann opened the pull request:
#### 6b5b349b0390030413d11288faf13659fd76ec5d
<pre>
[LBSE] Transform update handling is unreliable
<a href="https://bugs.webkit.org/show_bug.cgi?id=247671">https://bugs.webkit.org/show_bug.cgi?id=247671</a>

Reviewed by Rob Buis.

If a SVG renderer does not have a transformation and &lt;animateMotion&gt;
adds a non-identity supplementalTransform() no visual effect is observed.
Only when e.g. zooming in/out (-&gt; style invalidation) the animation starts
to be visible.

If e.g. the &apos;transform&apos; attribute changes SVG*Element::svgAttributeChanged()
needs to notify the renderer to re-evaluate the &apos;HasSVGTransform&apos; and
&apos;HasTransformRelatedProperty&apos; flags, since the renderer might have gained
or lost a transformation due to the attribute change. Therefore explicitely
call updateHasSVGTransformFlags() instead of either updateFromStyle() or
updateFromElement(). This avoids unnecessary extra work, and will allow us
to handle transformation changes without performing relayouts, soon.

SVGAnimateMotionElement was lacking similar logic to communicate
transformation changes -- the aforementioned flags were out of sync,
and thus only if the element was already transformed one could observe
the animation -- that&apos;s fixed now in general.

Revisit all renderers to enforce that updateFromStyle() is only called
from one place (RenderLayerModelObject::styleDidChange) except for one
exception: RenderSVGRoot needs to forward updateFromStyle() calls to
its anonymous child, the RenderSVGViewportContainer. RenderSVGRoot
no longer stores a pointer to the RenderSVGViewportContainer, but instead
grabs it from the render tree hierarchy (firstChild()). This allows to
ink the anonymous RenderSVGViewportContainer to its owning RenderSVGRoot,
simplifying the render tree building (see RenderTreeBuilderSVG).
updateFromElement() is no longer needed for SVG.

Transform change handling is now handled consistenly for CSS transforms,
SVG transforms, SMIL transforms, SVG &quot;generated&quot; / &quot;induced&quot; transforms
such as viewBox, x/y translation for &lt;use&gt; elements, etc. share the same logic.

Covered by existing tests - 25 tests are fixed by this patch in LBSE.

* LayoutTests/platform/mac-ventura-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/missing-xlink-expected.png: Removed.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/repaint/mask-clip-target-transform-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/zoom/page/zoom-hixie-mixed-008-expected.png:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/zoom/page/zoom-hixie-mixed-009-expected.png:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/zoom/page/zoom-hixie-rendering-model-004-expected.png: Removed.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/zoom/page/zoom-svg-float-border-padding-expected.png:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/zoom/text/zoom-hixie-mixed-008-expected.png: Removed.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/zoom/text/zoom-hixie-rendering-model-004-expected.png: Removed.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/zoom/text/zoom-svg-float-border-padding-expected.png: Removed.
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
(WebCore::RenderLayerModelObject::updateHasSVGTransformFlags):
* Source/WebCore/rendering/RenderLayerModelObject.h:
(WebCore::RenderLayerModelObject::needsHasSVGTransformFlags const):
(WebCore::RenderLayerModelObject::updateFromStyle):
* Source/WebCore/rendering/svg/RenderSVGBlock.cpp:
(WebCore::RenderSVGBlock::updateFromStyle):
(WebCore::RenderSVGBlock::needsHasSVGTransformFlags const):
* Source/WebCore/rendering/svg/RenderSVGBlock.h:
* Source/WebCore/rendering/svg/RenderSVGForeignObject.cpp:
(WebCore::RenderSVGForeignObject::updateFromStyle):
* Source/WebCore/rendering/svg/RenderSVGForeignObject.h:
* Source/WebCore/rendering/svg/RenderSVGHiddenContainer.h:
* Source/WebCore/rendering/svg/RenderSVGImage.cpp:
(WebCore::RenderSVGImage::needsHasSVGTransformFlags const):
* Source/WebCore/rendering/svg/RenderSVGImage.h:
* Source/WebCore/rendering/svg/RenderSVGInline.cpp:
(WebCore::RenderSVGInline::needsHasSVGTransformFlags const):
(WebCore::RenderSVGInline::updateFromStyle):
* Source/WebCore/rendering/svg/RenderSVGInline.h:
* Source/WebCore/rendering/svg/RenderSVGModelObject.cpp:
(WebCore::RenderSVGModelObject::updateFromStyle):
* Source/WebCore/rendering/svg/RenderSVGRoot.cpp:
(WebCore::RenderSVGRoot::viewportContainer const):
(WebCore::RenderSVGRoot::needsHasSVGTransformFlags const):
(WebCore::RenderSVGRoot::updateFromStyle):
(WebCore::RenderSVGRoot::setViewportContainer): Deleted.
(WebCore::RenderSVGRoot::updateFromElement): Deleted.
* Source/WebCore/rendering/svg/RenderSVGRoot.h:
* Source/WebCore/rendering/svg/RenderSVGShape.cpp:
(WebCore::RenderSVGShape::needsHasSVGTransformFlags const):
* Source/WebCore/rendering/svg/RenderSVGShape.h:
* Source/WebCore/rendering/svg/RenderSVGTransformableContainer.cpp:
(WebCore::RenderSVGTransformableContainer::needsHasSVGTransformFlags const):
(WebCore::RenderSVGTransformableContainer::updateFromStyle): Deleted.
* Source/WebCore/rendering/svg/RenderSVGTransformableContainer.h:
* Source/WebCore/rendering/svg/RenderSVGViewportContainer.cpp:
(WebCore::RenderSVGViewportContainer::RenderSVGViewportContainer):
(WebCore::RenderSVGViewportContainer::svgSVGElement const):
(WebCore::RenderSVGViewportContainer::needsHasSVGTransformFlags const):
(WebCore::RenderSVGViewportContainer::updateFromStyle):
(WebCore::RenderSVGViewportContainer::updateFromElement): Deleted.
(WebCore::RenderSVGViewportContainer::needsHasSVGTransformFlag const): Deleted.
* Source/WebCore/rendering/svg/RenderSVGViewportContainer.h:
* Source/WebCore/rendering/updating/RenderTreeBuilderSVG.cpp:
(WebCore::RenderTreeBuilder::SVG::attach):
(WebCore::RenderTreeBuilder::SVG::findOrCreateParentForChild):
* Source/WebCore/svg/SVGAnimateMotionElement.cpp:
(WebCore::SVGAnimateMotionElement::applyResultsToTarget):
* Source/WebCore/svg/SVGGraphicsElement.cpp:
(WebCore::SVGGraphicsElement::svgAttributeChanged):
* Source/WebCore/svg/SVGSVGElement.cpp:
(WebCore::SVGSVGElement::updateCurrentTranslate):
(WebCore::SVGSVGElement::svgAttributeChanged):
(WebCore::SVGSVGElement::didAttachRenderers): Deleted.
* Source/WebCore/svg/SVGSVGElement.h:

Canonical link: <a href="https://commits.webkit.org/256787@main">https://commits.webkit.org/256787@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c0caf1e455ab3599dbed9131b4d7d28e65290285

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96828 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6094 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29923 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106354 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/166636 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100806 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6310 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34823 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89210 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103054 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102502 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4726 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83440 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31689 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/86562 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/88431 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/74640 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/128 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19915 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/117 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21334 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4713 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/4909 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/43861 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1345 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40633 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

rwlbuis commented on Source/WebCore/rendering/svg/RenderSVGBlock.cpp:
```diff
@@ -40,6 +40,19 @@ RenderSVGBlock::RenderSVGBlock(SVGGraphicsElement& element, RenderStyle&& style)
+        return;
+
+    updateFromStyle();
+    updateLayerTransform();
```
Doesn't this mean updateFromStyle and updateLayerTransform get called twice for LBSE?

nikolaszimmermann replied:
Indeed - I could have mentioned this. The first call is nonsense, too early. styleDidChange() gets invoked just after the renderer received a style -- too early to compute SVG transform matrices. Due to the transform-box/transform-origin dependency the transform matrix of a container can only be fully known after children finished layout (and thus the container boundaries are well defined). However I agree, and it's good that you mention it, this should be addresses as well in this PR. I'll sleep about it :-)

nikolaszimmermann replied:
@rwlbuis you send me down the rabbit hole -- no more updateFromElement() calls, guaranteed one updateFromStyle() call per cycle, not multiple ones, only from one place, etc... much safer handling of the SVG transform flags, and most importantly: consistent for all LBSE renderers.

rwlbuis replied:
I like this one much better :)

rwlbuis reviewed (approved):
LGTM.

nikolaszimmermann committed:
[LBSE] Transform update handling is unreliable
https://bugs.webkit.org/show_bug.cgi?id=247671

Reviewed by Rob Buis.

If a SVG renderer does not have a transformation and <animateMotion>
adds a non-identity supplementalTransform() no visual effect is observed.
Only when e.g. zooming in/out (-> style invalidation) the animation starts
to be visible.

If e.g. the 'transform' attribute changes SVG*Element::svgAttributeChanged()
needs to notify the renderer to re-evaluate the 'HasSVGTransform' and
'HasTransformRelatedProperty' flags, since the renderer might have gained
or lost a transformation due to the attribute change. Therefore explicitely
call updateHasSVGTransformFlags() instead of either updateFromStyle() or
updateFromElement(). This avoids unnecessary extra work, and will allow us
to handle transformation changes without performing relayouts, soon.

SVGAnimateMotionElement was lacking similar logic to communicate
transformation changes -- the aforementioned flags were out of sync,
and thus only if the element was already transformed one could observe
the animation -- that's fixed now in general.

Revisit all renderers to enforce that updateFromStyle() is only called
from one place (RenderLayerModelObject::styleDidChange) except for one
exception: RenderSVGRoot needs to forward updateFromStyle() calls to
its anonymous child, the RenderSVGViewportContainer. RenderSVGRoot
no longer stores a pointer to the RenderSVGViewportContainer, but instead
grabs it from the render tree hierarchy (firstChild()). This allows to
ink the anonymous RenderSVGViewportContainer to its owning RenderSVGRoot,
simplifying the render tree building (see RenderTreeBuilderSVG).
updateFromElement() is no longer needed for SVG.

Transform change handling is now handled consistenly for CSS transforms,
SVG transforms, SMIL transforms, SVG "generated" / "induced" transforms
such as viewBox, x/y translation for <use> elements, etc. share the same logic.

Covered by existing tests - 25 tests are fixed by this patch in LBSE.

* LayoutTests/platform/mac-ventura-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/missing-xlink-expected.png: Removed.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/repaint/mask-clip-target-transform-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/zoom/page/zoom-hixie-mixed-008-expected.png:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/zoom/page/zoom-hixie-mixed-009-expected.png:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/zoom/page/zoom-hixie-rendering-model-004-expected.png: Removed.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/zoom/page/zoom-svg-float-border-padding-expected.png:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/zoom/text/zoom-hixie-mixed-008-expected.png: Removed.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/zoom/text/zoom-hixie-rendering-model-004-expected.png: Removed.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/zoom/text/zoom-svg-float-border-padding-expected.png: Removed.
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
(WebCore::RenderLayerModelObject::updateHasSVGTransformFlags):
* Source/WebCore/rendering/RenderLayerModelObject.h:
(WebCore::RenderLayerModelObject::needsHasSVGTransformFlags const):
(WebCore::RenderLayerModelObject::updateFromStyle):
* Source/WebCore/rendering/svg/RenderSVGBlock.cpp:
(WebCore::RenderSVGBlock::updateFromStyle):
(WebCore::RenderSVGBlock::needsHasSVGTransformFlags const):
* Source/WebCore/rendering/svg/RenderSVGBlock.h:
* Source/WebCore/rendering/svg/RenderSVGForeignObject.cpp:
(WebCore::RenderSVGForeignObject::updateFromStyle):
* Source/WebCore/rendering/svg/RenderSVGForeignObject.h:
* Source/WebCore/rendering/svg/RenderSVGHiddenContainer.h:
* Source/WebCore/rendering/svg/RenderSVGImage.cpp:
(WebCore::RenderSVGImage::needsHasSVGTransformFlags const):
* Source/WebCore/rendering/svg/RenderSVGImage.h:
* Source/WebCore/rendering/svg/RenderSVGInline.cpp:
(WebCore::RenderSVGInline::needsHasSVGTransformFlags const):
(WebCore::RenderSVGInline::updateFromStyle):
* Source/WebCore/rendering/svg/RenderSVGInline.h:
* Source/WebCore/rendering/svg/RenderSVGModelObject.cpp:
(WebCore::RenderSVGModelObject::updateFromStyle):
* Source/WebCore/rendering/svg/RenderSVGRoot.cpp:
(WebCore::RenderSVGRoot::viewportContainer const):
(WebCore::RenderSVGRoot::needsHasSVGTransformFlags const):
(WebCore::RenderSVGRoot::updateFromStyle):
(WebCore::RenderSVGRoot::setViewportContainer): Deleted.
(WebCore::RenderSVGRoot::updateFromElement): Deleted.
* Source/WebCore/rendering/svg/RenderSVGRoot.h:
* Source/WebCore/rendering/svg/RenderSVGShape.cpp:
(WebCore::RenderSVGShape::needsHasSVGTransformFlags const):
* Source/WebCore/rendering/svg/RenderSVGShape.h:
* Source/WebCore/rendering/svg/RenderSVGTransformableContainer.cpp:
(WebCore::RenderSVGTransformableContainer::needsHasSVGTransformFlags const):
(WebCore::RenderSVGTransformableContainer::updateFromStyle): Deleted.
* Source/WebCore/rendering/svg/RenderSVGTransformableContainer.h:
* Source/WebCore/rendering/svg/RenderSVGViewportContainer.cpp:
(WebCore::RenderSVGViewportContainer::RenderSVGViewportContainer):
(WebCore::RenderSVGViewportContainer::svgSVGElement const):
(WebCore::RenderSVGViewportContainer::needsHasSVGTransformFlags const):
(WebCore::RenderSVGViewportContainer::updateFromStyle):
(WebCore::RenderSVGViewportContainer::updateFromElement): Deleted.
(WebCore::RenderSVGViewportContainer::needsHasSVGTransformFlag const): Deleted.
* Source/WebCore/rendering/svg/RenderSVGViewportContainer.h:
* Source/WebCore/rendering/updating/RenderTreeBuilderSVG.cpp:
(WebCore::RenderTreeBuilder::SVG::attach):
(WebCore::RenderTreeBuilder::SVG::findOrCreateParentForChild):
* Source/WebCore/svg/SVGAnimateMotionElement.cpp:
(WebCore::SVGAnimateMotionElement::applyResultsToTarget):
* Source/WebCore/svg/SVGGraphicsElement.cpp:
(WebCore::SVGGraphicsElement::svgAttributeChanged):
* Source/WebCore/svg/SVGSVGElement.cpp:
(WebCore::SVGSVGElement::updateCurrentTranslate):
(WebCore::SVGSVGElement::svgAttributeChanged):
(WebCore::SVGSVGElement::didAttachRenderers): Deleted.
* Source/WebCore/svg/SVGSVGElement.h:

Canonical link: https://commits.webkit.org/256787@main
